### PR TITLE
adding TiledFusedLogitsLoss

### PR DESCRIPTION
**Pull request by stas00**
This PR adds `TiledFusedLogitsLoss` for an efficient fused logits+loss computation - this version pre-calculates grads in `forward`, avoiding recomputation in the backward.

**Please ignore for now. I'm waiting for Aurick to see if it works for SwiftKV before proceeding with reviews/merging.**

### CODE REVIEW

**Comment by stas00** [comment]
@tjruwase, could you please review?

it should be easy to review since it's a fork of an existing `SequenceTiledCompute` autograd class, just specialized for logit-loss, so it's easier to understand and invoke. i.e. nothing new.

**Review by sfc-gh-truwase — APPROVED** [review]

File: deepspeed/runtime/sequence_parallel/ulysses_sp.py [hunk 1]
```diff
@@ -885,6 +891,154 @@ def backward(ctx, *grads) -> torch.Tensor:
...
+            # if seqlen is not exactly divisible by shards the last step will be shorter than shard_step
+            shard_step = x_shards[i].shape[0]
+            shard_offset = i * x_shards[0].shape[0]
+
+            x_shard.grad = x_grad.narrow(0, shard_offset, shard_step).view_as(x_shard)
+
+            with torch.enable_grad():
+                args = (self, x_shard, y_shard)
+                if mask is not None:
+                    args.append(mask_shards[i])
```
Comment (sfc-gh-aqiao) [inline deepspeed/runtime/sequence_parallel/ulysses_sp.py:1015]: @stas00 this needs to be `args = args + (mask_shards[i],)`

Comment (sfc-gh-sbekman) [reply deepspeed/runtime/sequence_parallel/ulysses_sp.py:1015]: Thank you, Aurick - applying it here https://github.com/deepspeedai/DeepSpeed/pull/7459
